d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingState.java

```diff
@@ -0,0 +1,164 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.subsetting;
+
+import com.linkedin.d2.balancer.clients.TrackerClient;
+import java.net.URI;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.http.annotation.GuardedBy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * State of cluster subsetting
+ */
+public class SubsettingState
+{
+  private static final Logger LOG = LoggerFactory.getLogger(SubsettingState.class);
+  private final Object _lock = new Object();
+  private final SubsettingStrategyFactory _subsettingStrategyFactory;
+
+  /**
+   * Map from serviceName => Partition Id => subset map
+   */
+  @GuardedBy("_lock")
+  private final Map<String, Map<Integer, Map<URI, TrackerClient>>> _weightedSubsetsCache;
+
+  /**
+   * Map from serviceName => Partition Id => all potential clients
+   */
+  @GuardedBy("_lock")
+  private final Map<String, Map<Integer, Set<URI>>> _potentialClientsMap;
+
+  @GuardedBy("_lock")
+  private final Map<String, Integer> _minClusterSubsetSizeMap;
+
+  @GuardedBy("_lock")
+  private long _version;
```

Comment:
I didn't quite understand why do we need this version, can you explain? thx

Reply:
I refactored the logic in the SubsettingState in the latest commit. We now have a serviceName -> subsetCache map which I believe is a lot clearer. Please help take another look